File: backend/app/Models/Booking.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Booking extends Model
{
    protected $fillable = [
        'user_id',
        'kos_id',
        'start_date',
        'duration_months',
        'notes',
        'status',
        'admin_note',
        'payment_status',
        'paid_at',
    ];

    protected $casts = [
        'start_date' => 'date',
        'paid_at' => 'datetime',
    ];

    public function user()
    {
        return $this->belongsTo(User::class);
    }

    public function kos()
    {
        return $this->belongsTo(Kos::class);
    }

    /**
     * Dipakai buat memutuskan siapa yang BOLEH menulis ulasan baru --
     * sengaja "completed" saja (bukan "confirmed" juga), karena "confirmed"
     * cuma berarti pemilik menyetujui, belum tentu masa sewanya sudah
     * selesai/sungguh-sungguh sudah menginap. Dipakai bersama oleh
     * Api\ReviewController, Web\WebKosController, dan tempat lain yang
     * perlu tahu status ini (mis. tampilkan/sembunyikan tombol ulasan)
     * supaya syaratnya konsisten di satu tempat saja.
     */
    public static function userHasCompletedStayAt(int $userId, int $kosId): bool
    {
        return static::where('user_id', $userId)
            ->where('kos_id', $kosId)
            ->where('status', 'completed')
            ->exists();
    }
}

File: frontend/lib/models/booking.dart
import 'kos.dart';

class Booking {
  final int id;
  final int kosId;
  final Kos? kos;
  final int? userId;
  final String? userName;
  final DateTime startDate;
  final int durationMonths;
  final String? notes;
  final String status;
  final String? adminNote;
  final DateTime createdAt;
  final String paymentStatus;
  final DateTime? paidAt;

  Booking({
    required this.id,
    required this.kosId,
    this.kos,
    this.userId,
    this.userName,
    required this.startDate,
    required this.durationMonths,
    this.notes,
    required this.status,
    this.adminNote,
    required this.createdAt,
    this.paymentStatus = 'unpaid',
    this.paidAt,
  });

  bool get isPaid => paymentStatus == 'paid';

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['id'] ?? 0,
      kosId: json['kos_id'] ?? 0,
      kos: json['kos'] != null ? Kos.fromJson(json['kos']) : null,
      userId: json['user']?['id'] ?? json['user_id'],
      userName: json['user']?['name'],
      startDate: DateTime.parse(json['start_date']),
      durationMonths: json['duration_months'] ?? 1,
      notes: json['notes'],
      status: json['status'] ?? 'pending',
      adminNote: json['admin_note'],
      createdAt: DateTime.parse(json['created_at']),
      paymentStatus: json['payment_status'] ?? 'unpaid',
      paidAt: json['paid_at'] != null ? DateTime.tryParse(json['paid_at']) : null,
    );
  }
}
